src/Microsoft.Health.Fhir.SqlServer/Features/Schema/ResourceVersion.cs

```diff
@@ -0,0 +1,36 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using Microsoft.Data.SqlClient;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Schema
+{
+    public static class ResourceVersion
+    {
+        private static bool versionIsLong = true;
+
+        public static int GetVersion(SqlDataReader reader, int index)
+        {
+            int version = 0;
+            bool retry = true;
+
+            while (retry)
+            {
+                try
+                {
+                    version = versionIsLong ? (int)reader.GetInt64(index) : reader.GetInt32(index);
+                    retry = false;
+                }
+                catch (InvalidCastException)
+                {
+                    versionIsLong = !versionIsLong;
+                }
+            }
+
+            return version;
+        }
```

Comment:
I think @cunninghamjc mentioned this, if we check the type it seems like it should be as simple as an if/else? This seems like its worth testing, this is probably the most efficient implementation here.

```suggestion
        public static int GetVersion(SqlDataReader reader, int index)
        {
            if (reader.GetFieldType(index) == typeof(long))
            {
                return (int)reader.GetInt64(index);
            }
        
            return reader.GetInt32(index);
        }
```

Reply:
Please keep in mind that this is a throwaway code, so we should not be spending any extra time to make it super nice. 
BTW As far as efficiency, current code has single if based on single static bool and does not require checking reader metadata.

Reply:
The cyclomatic complexity of a single `if` statement is far lower than a `try`, a `ternary` , a `while` loop, a switching `bool`, you also have 2 local variable allocations.
You could cache the metadata check without any of the above also.

In the current implement if for some reason `version = versionIsLong ? (int)reader.GetInt64(index) : reader.GetInt32(index);` throws an `InvalidCastException` because it doesn't return an int or int64, its an infinate loop.

Even if its temporary work it, doesn't mean we should ignore all best practices. The very fact its added to handle an edge case and doesn't have a test feels a little uncomfortable.

Reply:
Does this approach also allow us to keep sequential access, since we wouldn't have to read the column twice?

Reply:
No, it does not. But this is irrelevant, as "read twice" happens once in the lifetime when schema is switched.

Reply:
That "In the current implement if for some reason version = versionIsLong ? (int)reader.GetInt64(index) : reader.GetInt32(index); throws an InvalidCastException because it doesn't return an int or int64, its an infinate loop." is a good point. We can ensure that we retry only once.

Reply:
On this "The cyclomatic complexity of a single if statement is far lower than a try, a ternary , a while loop, a switching bool, you also have 2 local variable allocations."
It is all true, but if we compare to reading compressed resource and unzipping it, it goes to a wash.

Reply:
Simplified the code reading version based on type.